src/app/services/todo-service.service.ts

```diff
@@ -0,0 +1,37 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { catchError, map, throwError } from 'rxjs';
+import { todos } from '../todo/todo';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TodoServiceService { 
+
+   url = "https://todo-backend-laravel.herokuapp.com/api";
+   
+  constructor(private http:HttpClient) { }
+
+  getAllTask(){
+    return this.http.get(this.url).pipe(
+      catchError((error:any)=> throwError(error))
+      )
+  }
+
+  postTask(data: todos){
+    return this.http.post(this.url,data).pipe(
+      catchError((error:any)=> throwError(error))
+      )
+  }
+  deleteTask(id: Number){
+    return this.http.delete(this.url+"/"+id).pipe(
+      catchError((error:any)=> throwError(error))
+      )
+  }
+
+  updateTask(id:number,data:any){
+    return this.http.patch(this.url+"/"+id,data).pipe(
+      catchError((error:any)=> throwError(error))
+      )
+  }
+}
```

Comment:
getAllTask and postTask can be combined into one function since the URL for both is same. Params can be a non-mandatory argument.
Similarly, deleteTask and updateTask can be combined into one.

Reply:
okay, I'm not aware of this but I'll try and explore it.